src/borg/fuse.py

```diff
@@ -35,22 +36,160 @@ def fuse_main():
+        # unices did this).
+        # The meta-array contains chunk IDs and item entries (described in iter_archive_items).
+        # The chunk IDs are referenced by item entries through relative offsets,
+        # which are bounded by the metadata chunk size.
+        self.meta = bytearray()
+        # The current write offset in self.meta
+        self.write_offset = 0
+
+        # Offset added to meta-indices, resulting in an inode,
+        # or substracted from inodes, resulting in a meta-indices.
+        # XXX: Merge FuseOperations.items and ItemCache to avoid this implicit limitation / hack
         self.offset = 1000000
 
-    def add(self, item):
-        pos = self.fd.seek(0, io.SEEK_END)
-        self.fd.write(msgpack.packb(item.as_dict()))
-        return pos + self.offset
+        # A temporary file that contains direct items, i.e. items directly cached in this layer.
+        # These are items that span more than one chunk and thus cannot be efficiently cached
+        # by the object cache (self.decrypted_repository), which would require variable-length structures;
+        # possible but not worth the effort, see iter_archive_items.
+        self.fd = tempfile.TemporaryFile(prefix='borg-tmp')
+
+        # A small LRU cache for chunks requested by ItemCache.get() from the object cache,
+        # this significantly speeds up directory traversal and similar operations which
+        # tend to re-read the same chunks over and over.
+        # The capacity is kept low because increasing it does not provide any significant advantage,
+        # but makes LRUCache's square behaviour noticeable as well as consuming some memory.
+        self.chunks = LRUCache(capacity=10, dispose=lambda _: None)
+
+        # Instrumentation
+        # Count of indirect items, i.e. data is cached in the object cache, not directly in this cache
+        self.indirect_items = 0
+        # Count of direct items, i.e. data is in self.fd
+        self.direct_items = 0
 
     def get(self, inode):
         offset = inode - self.offset
         if offset < 0:
             raise ValueError('ItemCache.get() called with an invalid inode number')
-        self.fd.seek(offset, io.SEEK_SET)
-        item = next(msgpack.Unpacker(self.fd, read_size=1024))
-        return Item(internal_dict=item)
+        if self.meta[offset] == ord(b'S'):
+            fd_offset = int.from_bytes(self.meta[offset + 1:offset + 9], 'little')
+            self.fd.seek(fd_offset, io.SEEK_SET)
+            return Item(internal_dict=next(msgpack.Unpacker(self.fd, read_size=1024)))
+        else:
+            _, chunk_id_relative_offset, chunk_offset = self.indirect_entry_struct.unpack_from(self.meta, offset)
+            chunk_id_offset = offset - chunk_id_relative_offset
+            # bytearray slices are bytearrays as well, explicitly convert to bytes()
+            chunk_id = bytes(self.meta[chunk_id_offset:chunk_id_offset + 32])
```

Comment:
> about speed: don't think it should matter that much and maybe we should not micro-optimize on that level.

No. ItemCache (before and after, sorting movements of data handling) is one of the biggest CPU time eaters when processing archives. In the current revision that's about 60%. These are at a point where you can remove a few VM ops, or replace expensive VM ops (such as member lookups) with cheaper ones (such as local variable lookups), like I did in iter_archive_bla with the meta and offset variables and get a tangible improvement.

Btw. I'd really prefer it if you would not quote me wrong. I referred to memory overhead, not throughput, although either is bad. (Python, and especially CPython, are simply not very good choices to achieve either, but that ship has sailed years ago for Borg of course... although I'll see in 1.2-dev what I can still salvage...).